web/packages/teleterm/src/services/fileStorage/fileStorage.ts

```diff
@@ -37,25 +37,37 @@ export interface FileStorage {
 
   /** Returns the file path used to create the storage. */
   getFilePath(): string;
+
+  /** Returns the error that could occur while reading and parsing the file. */
+  getFileLoadingError(): Error | undefined;
 }
 
-export function createFileStorage(opts: {
+export async function createFileStorage(opts: {
   filePath: string;
   debounceWrites: boolean;
-}): FileStorage {
+  /** Prevents state updates when the file has not been loaded correctly, so its content will not be overwritten. */
+  discardUpdatesWhenLoadingFileFailed?: boolean;
```

Comment:
I feel like this flag is where our past design decisions came back to bite us. 😶‍🌫️

Instead of adding a new flag, I feel like it'd be better to pass an alternative implementation of file storage that just ignores all writes and returns undefined for all reads. If you squint really hard, adding flags on top of each other is similar to using class inheritance to implement different behavior (remember that Sandi Metz video I linked to?).

`debounceWrites` as a flag just doesn't make any sense when `discardUpdatesWhenLoadingFileFailed` is set to true because then `debounceWrites` has no effect. Yet it's totally fine to pass `debounceWrites: true, discardUpdatesWhenLoadingFileFailed: true`. Instead of providing complex types to work our way around this, we could use a different implementation if there was an error.

Alas, we cannot really do that here as the error thrown during the file load is an integral part of a file storage and not a separate value that can be passed independently of the file storage that's being used.

Similar to my previous comment, I don't feel like it makes sense to refactor all of this now so I'm just leaving this comment for future consideration. Before we add a third flag or have to implement a similar object in the future, let's consider other solutions.

Reply:
I see your point, but are these flags really mutually exclusive?

>debounceWrites as a flag just doesn't make any sense when discardUpdatesWhenLoadingFileFailed is set to true because then debounceWrites has no effect

That's not entirely true. `debounceWrites` has no effect when `discardUpdatesWhenLoadingFileFailed` is set **and** an error occurred while loading the file. It will correctly debounce writes if the file has been loaded with no errors!

Reply:
You're right that they're indeed not mutually exclusive. Still, it was added to handle one specific situation which we could deal with differently with a different API.